Objects/unicodeobject.c

```diff
@@ -2986,45 +2989,38 @@ unicode_aswidechar(PyObject *unicode,
         return res + 1;
 }
 
-Py_ssize_t
-PyUnicode_AsWideChar(PyObject *unicode,
-                     wchar_t *w,
-                     Py_ssize_t size)
-{
-    if (unicode == NULL) {
-        PyErr_BadInternalCall();
-        return -1;
-    }
-    return unicode_aswidechar(unicode, w, size);
-}
-
 wchar_t*
 PyUnicode_AsWideCharString(PyObject *unicode,
                            Py_ssize_t *size)
 {
-    wchar_t* buffer;
+    const wchar_t *wstr;
+    wchar_t *buffer;
     Py_ssize_t buflen;
 
     if (unicode == NULL) {
         PyErr_BadInternalCall();
         return NULL;
     }
 
-    buflen = unicode_aswidechar(unicode, NULL, 0);
-    if (buflen == -1)
+    wstr = PyUnicode_AsUnicodeAndSize(unicode, &buflen);
+    if (wstr == NULL) {
         return NULL;
+    }
+    if (size == NULL && wcslen(wstr) != (size_t)buflen) {
+        PyErr_SetString(PyExc_ValueError,
+                        "embedded null character");
```

Comment:
This exception is for the users of extensions, not for the authors of extensions. `PyArg_ParseTuple()` raises the same exception, as well as some other functions. Using `PyArg_ParseTuple()` or `PyUnicode_AsWideCharString()` or other way for converting Python objects to C values is an implementation detail insignificant for the users of extensions. Words about a size parameter don't make sense at Python level.

The direct users of C API can read the `PyUnicode_AsWideCharString()` documentation when they get this exception (as well as they can read the `PyArg_ParseTuple()` documentation when they get the similar exception in `PyArg_ParseTuple()`). It is explicitly documented.